File: src/Core.Rust.TriBoolean/src/float.rs
//! Tri-boolean floating point -- biased-exponent decoder (B-0944 slice 5 pt2, Rust parity oracle #4).
//!
//! Mirrors the TS distribution surface (src/Core.TypeScript/tri-boolean-float) and the F#/C# ports
//! with the RATIFIED biased-exponent decoder. The middle field decodes the ends (middle-out,
//! self-describing):
//!
//! ```text
//! decoded value = V * 2 ^ (mode - bias),   bias = 2 ^ (decoderWidth - 1)
//! ```
//!
//! where `V` = MSB-first base-2 read of (high ++ low) and `mode` = MSB-first base-2 read of the
//! middle decoder field (`Tri::True` = 1, `Tri::False` = 0). Every field is a `Vec<Tri>` -- the
//! float is a composite of digital-qubit cells, so any trit may be held (`Tri::N`): `Tri::N` in a
//! value trit => the value is superposed; `Tri::N` in a decoder trit => the decode instruction is
//! superposed. `measure` is the only collapsing op; it surfaces which superposition is held as
//! `Err(FloatFeedback)` (Rust-native `Result`) rather than collapsing silently.
//!
//! rustc is oracle #4: a non-exhaustive `match` over `Tri` is a hard compile error, so the decode's
//! held-state handling is exhaustive by construction -- the strongest of the four oracles.

use crate::Tri;

/// Field widths of a tri-boolean float (trits per field), MSB-first within each field. Parity with
/// the TS `FloatShape`, F# `Float.FloatShape`, and C# `FloatShape`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct FloatShape {
    /// High (more-significant) value trits.
    pub high_width: usize,
    /// Middle decoder trits (the biased-exponent mode; read first).
    pub decoder_width: usize,
    /// Low (less-significant) value trits.
    pub low_width: usize,
}

impl FloatShape {
    /// Reference v0 shape: 4/3/4 -- 8 value trits, mode in `[0,8)`, bias = 4. Parity with C#
    /// `FloatShape.Default` / F# `defaultShape` / TS `DEFAULT_SHAPE`.
    pub const DEFAULT: Self = Self {
        high_width: 4,
        decoder_width: 3,
        low_width: 4,
    };
}

/// A tri-boolean float: a composite of digital-qubit cells. Each field is MSB-first; because every
/// position is a `Tri`, any trit may be held (`Tri::N`). Parity with the TS/F#/C# `TriFloat`.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct TriFloat {
    /// The field widths.
    pub shape: FloatShape,
    /// High value trits (MSB-first).
    pub high: Vec<Tri>,
    /// Middle decoder trits (the biased-exponent mode, MSB-first).
    pub decoder: Vec<Tri>,
    /// Low value trits (MSB-first).
    pub low: Vec<Tri>,
}

/// Which superposition is held when [`measure`] cannot collapse a float to a single number. The two
/// held-states are distinct. Parity with the F# `Float.FloatFeedback` DU / TS `FloatFeedback` union.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum FloatFeedback {
    /// `Tri::N` in the decoder field -- the decode instruction itself is superposed.
    InterpretationSuperposed,
    /// `Tri::N` in a value trit while the decoder is certain -- the value is superposed.
    ValueSuperposed,
}

/// MSB-first base-2 read of a trit field (`Tri::True` = 1, `Tri::False` = 0). `None` iff ANY trit is
/// held (`Tri::N`) -- the held signal. Exhaustive `match` (no fourth state; Rust has no null).
fn int_of<'a>(trits: impl Iterator<Item = &'a Tri>) -> Option<u64> {
    let mut v: u64 = 0;
    for t in trits {
        match t {
            Tri::N => return None,
            Tri::True => v = v * 2 + 1,
            Tri::False => v *= 2,
        }
    }
    Some(v)
}

/// MSB-first encode of a non-negative integer into `width` certain trits.
fn int_to_trits(v: u64, width: usize) -> Vec<Tri> {
    (0..width)
        .map(|i| {
            let shift = (width - 1 - i) as u32;
            if ((v >> shift) & 1) == 1 {
                Tri::True
            } else {
                Tri::False
            }
        })
        .collect()
}

/// decode (middle-out, biased-exponent): read the MIDDLE decoder first, then decode OUTWARD.
/// `Tri::N` in the decoder => [`FloatFeedback::InterpretationSuperposed`]; `Tri::N` in a value trit
/// => [`FloatFeedback::ValueSuperposed`] (decoder read first, so interpretation dominates when both
/// are held); else `Ok(V * 2 ^ (mode - bias))`, bias = `2 ^ (decoderWidth - 1)`.
///
/// # Errors
/// Returns [`FloatFeedback::InterpretationSuperposed`] or [`FloatFeedback::ValueSuperposed`] when a
/// decoder or value trit (respectively) is held.
pub fn decode(f: &TriFloat) -> Result<f64, FloatFeedback> {
    let mode = match int_of(f.decoder.iter()) {
        None => return Err(FloatFeedback::InterpretationSuperposed),
        Some(m) => m,
    };
    let v = match int_of(f.high.iter().chain(f.low.iter())) {
        None => return Err(FloatFeedback::ValueSuperposed),
        Some(v) => v,
    };
    let bias = 1i32 << (f.decoder.len() as u32 - 1);
    let exp = mode as i32 - bias;
    Ok(v as f64 * 2f64.powi(exp))
}

/// measure: the only collapsing op (identical to [`decode`]; named for parity with the cell's
/// `measure`/`cooperate` pair).
///
/// # Errors
/// Same as [`decode`].
pub fn measure(f: &TriFloat) -> Result<f64, FloatFeedback> {
    decode(f)
}

/// cooperate: engage WITHOUT collapsing -- identity, preserving every held (`Tri::N`) trit. Takes
/// the float by value and returns it unchanged (the move-through identity; no clone, nothing lost).
#[must_use]
pub fn cooperate(f: TriFloat) -> TriFloat {
    f
}

/// True iff the float cannot collapse to a single number (any value or decoder trit is held).
#[must_use]
pub fn is_held(f: &TriFloat) -> bool {
    decode(f).is_err()
}

/// Construct a float directly from trit fields (tests / advanced use); the shape is inferred from
/// the field lengths.
#[must_use]
pub fn from_trits(high: Vec<Tri>, decoder: Vec<Tri>, low: Vec<Tri>) -> TriFloat {
    let shape = FloatShape {
        high_width: high.len(),
        decoder_width: decoder.len(),
        low_width: low.len(),
    };
    TriFloat {
        shape,
        high,
        decoder,
        low,
    }
}

/// fromValue (biased-exponent canonical encode): find a `(mode, V)` with `V * 2 ^ (mode - bias) =
/// value`, `V` a non-negative integer fitting the value field and `mode` in the decoder field. Picks
/// the SMALLEST mode that works (a canonical representation; the biased-exponent decoder has
/// redundant representations). v0 is unsigned + finite. Round-trips with [`decode`].
///
/// # Errors
/// Returns a reason string when `value` is negative / non-finite, or when no `(mode, V)` represents
/// it in the given shape.
pub fn from_value(value: f64, shape: FloatShape) -> Result<TriFloat, String> {
    if value.is_nan() || value.is_infinite() || value < 0.0 {
        return Err("v0 is unsigned + finite".to_string());
    }
    let value_bits = shape.high_width + shape.low_width;
    let max_mode = (1u64 << shape.decoder_width) - 1;
    let max_v = 1u64 << value_bits;
    let bias = 1i32 << (shape.decoder_width as u32 - 1);
    for mode in 0..=max_mode {
        let exp = mode as i32 - bias;
        let scaled = value / 2f64.powi(exp); // = V
        if scaled.fract() == 0.0 && scaled >= 0.0 && scaled < max_v as f64 {
            let v = scaled as u64;
            let bits = int_to_trits(v, value_bits);
            return Ok(TriFloat {
                shape,
                high: bits[..shape.high_width].to_vec(),
                decoder: int_to_trits(mode, shape.decoder_width),
                low: bits[shape.high_width..].to_vec(),
            });
        }
    }
    Err(format!(
        "no (mode,V) with mode<={max_mode} and V<{max_v} represents {value}"
    ))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Tri::{False as F, N, True as T};

    // Rust parity oracle (#4 of four) for the biased-exponent tri-boolean float (B-0944 slice 5
    // pt2). Vectors mirror tests/Tests.FSharp/TriBoolean/Float.Tests.fs +
    // tests/Tests.CSharp/TriBoolean/FloatTests.cs so four-of-four parity IS the summonable-BFT
    // ballot. Shape 4/3/4: decoderWidth 3 -> bias 4; valueBits 8 -> V in [0,256).

    fn mk(high: Vec<Tri>, decoder: Vec<Tri>, low: Vec<Tri>) -> TriFloat {
        from_trits(high, decoder, low)
    }

    fn zero4() -> Vec<Tri> {
        vec![F, F, F, F]
    }

    #[test]
    fn decode_mode_equals_bias_value_is_v() {
        // decoder 100 = mode 4 = bias -> exp 0; low 0101 -> V = 5 -> 5.0
        let f = mk(zero4(), vec![T, F, F], vec![F, T, F, T]);
        assert_eq!(Ok(5.0), decode(&f));
    }

    #[test]
    fn decode_mode_greater_than_bias_times_two() {
        // decoder 101 = mode 5 -> exp +1; low 0011 -> V = 3 -> 6.0
        let f = mk(zero4(), vec![T, F, T], vec![F, F, T, T]);
        assert_eq!(Ok(6.0), decode(&f));
    }

    #[test]
    fn decode_mode_less_than_bias_half() {
        // decoder 011 = mode 3 -> exp -1; low 1000 -> V = 8 -> 4.0
        let f = mk(zero4(), vec![F, T, T], vec![T, F, F, F]);
        assert_eq!(Ok(4.0), decode(&f));
    }

    #[test]
    fn decode_mode_less_than_bias_quarter() {
        // decoder 010 = mode 2 -> exp -2; low 0100 -> V = 4 -> 1.0
        let f = mk(zero4(), vec![F, T, F], vec![F, T, F, F]);
        assert_eq!(Ok(1.0), decode(&f));
    }

    #[test]
    fn decode_reads_high_then_low_as_one_v() {
        // high 0001, low 0000 -> V = 0b00010000 = 16; decoder 100 = mode 4 -> exp 0 -> 16.0
        let f = mk(vec![F, F, F, T], vec![T, F, F], zero4());
        assert_eq!(Ok(16.0), decode(&f));
    }

    #[test]
    fn n_in_decoder_is_interpretation_superposed() {
        let f = mk(zero4(), vec![T, N, F], vec![F, T, F, T]);
        assert_eq!(Err(FloatFeedback::InterpretationSuperposed), decode(&f));
    }

    #[test]
    fn n_in_value_decoder_certain_is_value_superposed() {
        let f = mk(vec![F, F, F, N], vec![T, F, F], zero4());
        assert_eq!(Err(FloatFeedback::ValueSuperposed), decode(&f));
    }

    #[test]
    fn decoder_read_first_both_held_is_interpretation_superposed() {
        // N in BOTH decoder and value -> InterpretationSuperposed dominates (decoder checked first).
        let f = mk(vec![N, F, F, F], vec![N, F, F], zero4());
        assert_eq!(Err(FloatFeedback::InterpretationSuperposed), decode(&f));
    }

    #[test]
    fn measure_equals_decode() {
        let f = mk(zero4(), vec![T, F, F], vec![F, T, F, T]);
        assert_eq!(decode(&f), measure(&f));
    }

    #[test]
    fn cooperate_is_identity_and_preserves_held_trits() {
        let held = mk(vec![F, F, F, N], vec![N, F, F], zero4());
        assert_eq!(held.clone(), cooperate(held));
    }

    #[test]
    fn is_held_true_iff_any_value_or_decoder_trit_is_held() {
        let certain = mk(zero4(), vec![T, F, F], vec![F, T, F, T]);
        let held = mk(vec![F, F, F, N], vec![T, F, F], zero4());
        assert!(!is_held(&certain));
        assert!(is_held(&held));
    }

    #[test]
    fn from_value_round_trips_through_decode() {
        for v in [0.0, 1.0, 5.0, 6.0, 0.5, 8.0, 16.0] {
            let f = from_value(v, FloatShape::DEFAULT)
                .unwrap_or_else(|e| panic!("from_value {v} unexpectedly failed: {e}"));
            assert_eq!(Ok(v), decode(&f));
        }
    }

    #[test]
    fn from_value_surfaces_feedback_for_negative_and_non_dyadic() {
        assert!(from_value(-1.0, FloatShape::DEFAULT).is_err());
        assert!(from_value(0.1, FloatShape::DEFAULT).is_err()); // 0.1 is not a dyadic rational
    }
}

File: src/Core.FSharp.TriBoolean/Float.fs
namespace Zeta.Core.FSharp.TriBoolean

/// Tri-boolean floating point -- biased-exponent decoder (081KSV2WD0008QG0R00051XS0N slice 5 pt2, F# parity oracle).
///
/// Mirrors the TS distribution surface (src/Core.TypeScript/tri-boolean-float) with the RATIFIED
/// biased-exponent decoder. The middle field decodes the ends (middle-out, self-describing):
///
///   decoded value = V * 2 ^ (mode - bias),   bias = 2 ^ (decoderWidth - 1)
///
/// where V    = MSB-first base-2 read of (high ++ low)   (Tri.T = 1, Tri.F = 0), and
///       mode = MSB-first base-2 read of the middle decoder field.
///
/// Every field is a `Tri list` -- the float is a composite of digital-qubit cells (Types.fs), so
/// any trit may be held (Tri.N):
///   * Tri.N in a value trit   => the value is superposed       (ValueSuperposed)
///   * Tri.N in a decoder trit => the decode instruction itself  (InterpretationSuperposed)
/// `measure` is the only collapsing op; it surfaces which superposition is held rather than
/// collapsing silently (the digital-qubit measure/cooperate discipline at the number scope).
///
/// Width: field reads accumulate into int64 (matching the Rust u64 oracle; TS uses f64). This keeps
/// all four oracles in agreement up to f64's 2^53 exact-integer range -- the shared limit, since the
/// decoded value is f64 in every oracle. (A naive 32-bit int would wrap at 2^31, diverging from the
/// other three oracles before the f64 limit; int64 defers that to widths no realistic shape reaches.)
///
/// F# is oracle #2 of four (TS/F#/C#/Rust) in the summonable-BFT cross-language consensus. The F#
/// compiler is a non-Byzantine oracle: it cannot lie about whether the TS design type-composes, so
/// four-of-four compiler-checked parity = BFT consensus with no human voters.
module Float =

    /// Field widths (trits per field), MSB-first within each field.
    type FloatShape =
        { HighWidth: int
          DecoderWidth: int
          LowWidth: int }

    /// Reference v0 shape: 4/3/4 -- 8 value trits, mode in [0,8), bias = 4.
    let defaultShape: FloatShape =
        { HighWidth = 4
          DecoderWidth = 3
          LowWidth = 4 }

    /// A tri-boolean float: a composite of digital-qubit cells. Each field is MSB-first; any trit
    /// may be held (Tri.N).
    type TriFloat =
        { Shape: FloatShape
          High: Tri list
          Decoder: Tri list
          Low: Tri list }

    /// Which superposition is held when `measure` cannot collapse to a single number. The two
    /// held-states are distinct (the decode instruction is held vs the value is held).
    [<RequireQualifiedAccess>]
    type FloatFeedback =
        /// Tri.N in the decoder field -- the decode instruction itself is superposed.
        | InterpretationSuperposed
        /// Tri.N in a value trit while the decoder is certain -- the value is superposed.
        | ValueSuperposed

    /// MSB-first base-2 read of a trit field (Tri.T = 1, Tri.F = 0). None if ANY trit is held (Tri.N).
    /// Accumulates into int64 (see module doc): no wrap until 63 bits, well past any realistic shape
    /// and past f64's 2^53 exact range that bounds the decoded value across all four oracles.
    let private intOf (trits: Tri list) : int64 option =
        let rec go acc ts =
            match ts with
            | [] -> Some acc
            | Tri.N :: _ -> None
            | Tri.T :: rest -> go (acc * 2L + 1L) rest
            | Tri.F :: rest -> go (acc * 2L) rest

        go 0L trits

    /// MSB-first encode of a non-negative integer into `width` certain trits (Tri.T = 1, Tri.F = 0).
    let private intToTrits (v: int64) (width: int) : Tri list =
        [ for i in (width - 1) .. -1 .. 0 -> if (v >>> i) &&& 1L = 1L then Tri.T else Tri.F ]

    /// decode (middle-out, biased-exponent): read the MIDDLE decoder first, then decode OUTWARD.
    ///   * Tri.N in the decoder  => InterpretationSuperposed (the decode instruction is held).
    ///   * Tri.N in a value trit => ValueSuperposed (the value is held; interpretation known).
    ///   * else                  => Ok (V * 2 ^ (mode - bias)),  bias = 2 ^ (decoderWidth - 1).
    /// The decoder is read first, so InterpretationSuperposed dominates when both are held.
    let decode (f: TriFloat) : Result<float, FloatFeedback> =
        match intOf f.Decoder with
        | None -> Error FloatFeedback.InterpretationSuperposed
        | Some mode ->
            match intOf (f.High @ f.Low) with
            | None -> Error FloatFeedback.ValueSuperposed
            | Some v ->
                let bias = pown 2L (f.Decoder.Length - 1)
                Ok(float v * (2.0 ** float (mode - bias)))

    /// measure: the ONLY collapsing operation (identical to decode; named for parity with the
    /// digital-qubit cell primitive's measure/cooperate pair).
    let measure (f: TriFloat) : Result<float, FloatFeedback> = decode f

    /// cooperate: engage WITHOUT collapsing -- identity, preserving every held (Tri.N) trit. The
    /// wonder-compression-safe operation at the number scope.
    let cooperate (f: TriFloat) : TriFloat = f

    /// True iff the float cannot be collapsed to a single number (any value or decoder trit is held).
    let isHeld (f: TriFloat) : bool =
        match decode f with
        | Ok _ -> false
        | Error _ -> true

    /// Construct a float directly from trit fields (tests / advanced use). The shape is inferred
    /// from the field lengths.
    let fromTrits (high: Tri list) (decoder: Tri list) (low: Tri list) : TriFloat =
        { Shape =
            { HighWidth = List.length high
              DecoderWidth = List.length decoder
              LowWidth = List.length low }
          High = high
          Decoder = decoder
          Low = low }

    /// fromValue (biased-exponent canonical encode): find a (mode, V) with
    /// `V * 2 ^ (mode - bias) = value`, V a non-negative integer that fits the value field and mode
    /// in the decoder field. Picks the SMALLEST mode that works (a canonical representation; the
    /// biased-exponent decoder has redundant representations, so a canonical choice is required).
    /// Surfaces a reason string when not representable. v0 is unsigned + finite. Round-trips with
    /// decode: `fromValue v shape |> Result.map decode` resolves back to Ok v. Bounds math uses
    /// int64 so wide shapes return feedback rather than overflowing to negative/incorrect bounds.
    let fromValue (value: float) (shape: FloatShape) : Result<TriFloat, string> =
        if System.Double.IsNaN value || System.Double.IsInfinity value || value < 0.0 then
            Error "v0 is unsigned + finite"
        else
            let valueBits = shape.HighWidth + shape.LowWidth
            let maxMode = if shape.DecoderWidth >= 62 then System.Int64.MaxValue else (1L <<< shape.DecoderWidth) - 1L
            let maxV = if valueBits >= 62 then System.Int64.MaxValue else pown 2L valueBits
            let bias = if shape.DecoderWidth >= 62 then System.Int64.MaxValue / 2L else pown 2L (shape.DecoderWidth - 1)

            if value = 0.0 then
                let bits = List.init valueBits (fun _ -> Tri.F)
                Ok { Shape = shape
                     High = bits |> List.take shape.HighWidth
                     Decoder = intToTrits 0L shape.DecoderWidth
                     Low = bits |> List.skip shape.HighWidth }
            else
                let log2v = System.Math.Log2 value
                let startMode = int64 (float bias - float valueBits + System.Math.Floor(log2v) - 2.0)
                let rec tryMode (mode: int64) =
                    if mode > maxMode then
                        Error(sprintf "no (mode,V) with mode<=%d and V<%d represents %g" maxMode maxV value)
                    else
                        let scaled = value / (2.0 ** float (mode - bias))
                        if scaled < 1.0 then
                            Error(sprintf "no (mode,V) with mode<=%d and V<%d represents %g" maxMode maxV value)
                        elif System.Double.IsInteger scaled && scaled >= 0.0 && scaled < float maxV then
                            let v = int64 scaled
                            let bits = intToTrits v valueBits
                            Ok { Shape = shape
                                 High = bits |> List.take shape.HighWidth
                                 Decoder = intToTrits mode shape.DecoderWidth
                                 Low = bits |> List.skip shape.HighWidth }
                        else
                            tryMode (mode + 1L)

                tryMode (max 0L startMode)
